# ğŸª‘grayscale_ğŸ¬kifuwarabe_rpg
#	ãƒ‡ã‚£ãƒ¬ã‚¯ã‚¿ãƒ¼ï¼ˆğŸŒDirectorï¼›ç›£ç£ï¼‰
#	ã¨ã‚Šã‚ãˆãšã€ã‚²ãƒ¼ãƒ å…¨ä½“ã‚’ã‚³ãƒ³ãƒˆãƒ­ãƒ¼ãƒ«ã™ã‚‹
extends Node2D


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãƒ¡ãƒ¢ãƒªé–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# çŠ¶æ…‹ã€‚ WaitForKeyConfig, KeyConfig, Ready, Main ã®ï¼”ã¤
var current_state = &"WaitForKeyConfig"


# ç¾åœ¨é³´ã£ã¦ã„ã‚‹èƒŒæ™¯éŸ³æ¥½ã®ãƒãƒ¼ãƒ‰å
var current_bg_music_name = null
# ç¾åœ¨é³´ã£ã¦ã„ã‚‹åŠ¹æœéŸ³ã®ãƒãƒ¼ãƒ‰å
var current_se_name = null

# ãƒˆæ›¸ãï¼ˆã‚·ãƒŠãƒªã‚ªã®å‘½ä»¤ãƒ‘ãƒ©ã‚°ãƒ©ãƒ•ï¼‰ã§ä½¿ãˆã‚‹å¤‰æ•°ã®è¾æ›¸
var stage_directions_variables = {}
# ç–‘ä¼¼çš„ãªã‚¹ãƒªãƒ¼ãƒ—ã«ä½¿ã†ã‚«ã‚¦ãƒ³ãƒˆ
var sleep_seconds = 0.0


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãƒãƒ¼ãƒ‰ãƒ»ãƒ‘ã‚¹é–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# ãƒ‡ã‚£ãƒ¬ã‚¯ã‚¿ãƒ¼ãƒ»ãƒãƒ–å–å¾—
func monkey():
	return $"ğŸµMonkey"


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# åˆæœŸåŒ–
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# ã‚µãƒ–ãƒ„ãƒªãƒ¼ãŒå…¨ã¦ã‚¤ãƒ³ã‚¹ã‚¿ãƒ³ã‚¹åŒ–ã•ã‚ŒãŸã¨ãã«å‘¼ã³å‡ºã•ã‚Œã‚‹
# Called when the node enters the scene tree for the first time.
func _ready():

	# ã‚­ãƒ£ãƒƒã‚·ãƒ¥ã‚’ä½œæˆã™ã‚‹ã ã‘
	var _all_instruction_code = self.monkey().programmer().scenario_player().get_all_instruction_codes()

	# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
	# éè¡¨ç¤º
	# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
	#
	# é–‹ç™ºä¸­ã«ã„ã˜ã£ãŸã‚‚ã®ãŒæ®‹ã£ã¦ã‚‹ã‹ã‚‚ã—ã‚Œãªã„ã‹ã‚‰ã€æƒé™¤
	
	# ã‚°ãƒªãƒƒãƒ‰ã¯éš ã™
	self.monkey().grid().hide()
	
	# ã‚¤ãƒ©ã‚¹ãƒˆãƒ¬ãƒ¼ã‚¿ãƒ¼ã¯ã¨ã«ã‹ãéš ã™
	self.search_in_folder(
			self.monkey().illustrator(),		# æ¢ã™å ´æ‰€
			func(child_node):
				return child_node is Sprite2D,
			func(child_node):
				child_node.hide())
	
	# ãƒ†ãƒ­ãƒƒãƒ—ã¯ã¨ã«ã‹ãéè¡¨ç¤ºã«ã™ã‚‹
	self.search_in_folder(
			self.monkey().telop_coordinator(),		# æ¢ã™å ´æ‰€
			func(child_node):
				return child_node is CanvasLayer,
			func(child_node):
				child_node.hide())

	# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
	# è¡¨ç¤º
	# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
	
	# ç›£ç£è‡ªèº«
	self.show()
	# ã‚¤ãƒ©ã‚¹ãƒˆãƒ¬ãƒ¼ã‚¿ãƒ¼
	self.monkey().illustrator().show()
	# ãƒ†ãƒ­ãƒƒãƒ—
	self.monkey().telop_coordinator().show()


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãã®ä»–
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼


# TODO ãƒ©ã‚¤ãƒ–ãƒ©ãƒªã¸ç§»å‹•ã—ãŸã„
func search_in_folder(
		folder,				# æ¢ã™å ´æ‰€
		is_match,
		on_node_found):
	for child_node in folder.get_children():
		if child_node.name.begins_with("ğŸ“‚"):
			self.search_in_folder(
					child_node,
					is_match,
					on_node_found)

		elif is_match.call(child_node):
			on_node_found.call(child_node)


func on_key_config_entered():
	# èƒŒæ™¯
	self.monkey().programmer().images.find_node("ğŸ—»å´å·é§…å‰").visible = true


func on_key_config_exited():
	self.current_state = &"Ready"


func _process(delta):

	# ã‚­ãƒ¼ãƒ»ã‚³ãƒ³ãƒ•ã‚£ã‚°ãŒå§‹ã¾ã‚‹
	if self.current_state == &"WaitForKeyConfig":
		self.monkey().key_config_island().entry()
		self.current_state = &"KeyConfig"

	# ã‚­ãƒ¼ãƒ»ã‚³ãƒ³ãƒ•ã‚£ã‚°ã«åˆ¶å¾¡ã‚’è­²ã‚‹
	elif self.current_state == &"KeyConfig":
		self.monkey().key_config_island().on_process(delta)

	# ä¸»ãªçŠ¶æ…‹ã®å‰ã«
	elif self.current_state == &"Ready":
		self.current_state = &"Main"
		# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
		# æº–å‚™
		# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼

		# æœ€åˆã«å®Ÿè¡Œã™ã‚‹éƒ¨é–€å
		self.monkey().programmer().current_department_name = self.monkey().scenario_writer().department_control().start_department_name

		# ãƒ‘ãƒ¼ã‚¹ã™ã‚‹ãª
		self.monkey().programmer().scenario_player().get_current_department_value().set_parse_lock(true)

		# å°æœ¬ã®ã€ŒÂ§ã€ã‚»ã‚¯ã‚·ãƒ§ãƒ³ã®å†ç”Ÿ
		self.monkey().programmer().scenario_player().play_section()

		# ä¼è¨€çª“ã‚’ã€ä¸€æ™‚çš„ã«å±…ãªããªã£ã¦ã„ãŸã®ã‚’è§£é™¤ã™ã‚‹
		self.monkey().programmer().scenario_player().get_current_message_window_gui().set_appear_subtree(true)

	# ä¸»ãªçŠ¶æ…‹ã«åˆ¶å¾¡ã‚’è­²ã‚‹
	elif self.current_state == &"Main":
		self.monkey().programmer().scenario_player().on_process(delta)


# ãƒ†ã‚­ã‚¹ãƒˆãƒœãƒƒã‚¯ã‚¹ãªã©ã«ãƒ•ã‚©ãƒ¼ã‚«ã‚¹ãŒç„¡ã„ã¨ãã®ã‚­ãƒ¼å…¥åŠ›ã‚’æ‹¾ã†
#
# å­è¦ç´ ã‹ã‚‰è¦ªè¦ç´ ã®é †ã§å‘¼ã³å‡ºã•ã‚Œã‚‹ã‚ˆã†ã ã€‚
# ã“ã®ãƒ—ãƒ­ã‚°ãƒ©ãƒ ã§ã¯ã€€ãƒ«ãƒ¼ãƒˆã€€ã ã‘ã§ã€€ã‚­ãƒ¼å…¥åŠ›ã‚’æ‹¾ã†ã“ã¨ã«ã™ã‚‹
func _unhandled_key_input(event):

	# ã‚­ãƒ¼ãƒ»ã‚³ãƒ³ãƒ•ã‚£ã‚°ã®ãŸã‚ã«ã€ä½•ã‚‚ã™ã‚‹ãª
	if self.current_state == &"WaitForKeyConfig":
		pass

	# ã‚­ãƒ¼ãƒ»ã‚³ãƒ³ãƒ•ã‚£ã‚°ä¸­ãªã®ã§ã€ä½•ã‚‚ã™ã‚‹ãª
	elif self.current_state == &"KeyConfig":
		pass

	# ä¸»è¦ãªçŠ¶æ…‹
	elif self.current_state == &"Main":

		var vk_operation = null

		# ä½•ã‹ã‚­ãƒ¼ã‚’æŠ¼ã—ãŸã¨ã
		if event.is_pressed():
			print("ï¼»ç›£ç£ï¼½ã€€ã‚­ãƒ¼å…¥åŠ›ã€€æŠ¼ä¸‹")
			vk_operation = &"VKO_Pressed"
		
		# ä½•ã‹ã‚­ãƒ¼ã‚’é›¢ã—ãŸã¨ã
		elif event.is_released():
			print("ï¼»ç›£ç£ï¼½ã€€ã‚­ãƒ¼å…¥åŠ›ã€€ãƒªãƒªãƒ¼ã‚¹")
			vk_operation = &"VKO_Released"
		
		# ãã‚Œä»¥å¤–ã«ã¯å¯¾å¿œã—ã¦ãªã„
		else:
			print("ï¼»ç›£ç£ï¼½ã€€ã‚­ãƒ¼å…¥åŠ›ã€€â–²ï¼æƒ³å®šå¤–")
			return

		# ä»¥ä¸‹ã€ä»®æƒ³ã‚­ãƒ¼

		# ã“ã®ã‚²ãƒ¼ãƒ ç‹¬è‡ªã®ä»®æƒ³ã‚­ãƒ¼ã«å¤‰æ›
		var virtual_key_name = null
		
		# ã‚¨ãƒ³ã‚¿ãƒ¼ã‚­ãƒ¼æŠ¼ä¸‹
		if event.keycode == KEY_ENTER:
			virtual_key_name = &"VK_Ok"

		# ã‚¨ã‚¹ã‚±ãƒ¼ãƒ—ã‚­ãƒ¼æŠ¼ä¸‹
		elif event.keycode == KEY_ESCAPE:
			virtual_key_name = &"VK_Cancel"

		# ï¼»ï¼²ï¼½ã‚­ãƒ¼æŠ¼ä¸‹ï¼ˆå¾Œã§ã‚¹ãƒ¼ãƒ‘ãƒ¼ãƒ•ã‚¡ãƒŸã‚³ãƒ³ã® R ã‚­ãƒ¼ã«ã—ã‚ˆã†ã¨æ€ã£ã¦ã„ãŸã‚¢ãƒ«ãƒ•ã‚¡ãƒ™ãƒƒãƒˆï¼‰
		elif event.keycode == KEY_R:
			virtual_key_name = &"VK_FastForward"
		
		# ãã‚Œä»¥å¤–ã®ã‚­ãƒ¼ã¯ç„¡è¦–ã™ã‚‹ï¼ˆåå­—ã‚­ãƒ¼ã‚„ Ctrl ã‚­ãƒ¼ã®åˆ¤å®šã‚’å–ã‚Šé™¤ãã®ãŒé›£ã—ã„ï¼‰
		else:
			return

		var lever_value = 0.0

		# ä»®æƒ³ã‚­ãƒ¼ã‚’æŠ¼ä¸‹ã—ãŸã¨ã„ã†å»ºä»˜ã‘
		self.on_virtual_key_input(virtual_key_name, lever_value, vk_operation)


# ãƒ†ã‚­ã‚¹ãƒˆãƒœãƒƒã‚¯ã‚¹ãªã©ã«ãƒ•ã‚©ãƒ¼ã‚«ã‚¹ãŒç„¡ã„ã¨ãã®å…¥åŠ›ã‚’ã¨ã«ã‹ãæ‹¾ã†
func _unhandled_input(event):

	# ã‚­ãƒ¼ãƒ»ã‚³ãƒ³ãƒ•ã‚£ã‚°ã®ãŸã‚ã«ä½•ã‚‚ã™ã‚‹ãªã€ã¨ã„ã†çŠ¶æ…‹
	if self.current_state == &"WaitForKeyConfig":
		pass

	# ã‚­ãƒ¼ãƒ»ã‚³ãƒ³ãƒ•ã‚£ã‚°ã«å…¥åŠ›ã®åˆ¶å¾¡ã‚’è­²ã‚Œã€ã¨ã„ã†çŠ¶æ…‹
	elif self.current_state == &"KeyConfig":
		self.monkey().key_config_island().on_unhandled_input(event)

	# ä¸»ãªçŠ¶æ…‹
	elif self.current_state == &"Main":

		var vk_operation = null

		# ä½•ã‹ã‚­ãƒ¼ã‚’æŠ¼ã—ãŸã¨ã
		if event.is_pressed():
			#print("ï¼»ç›£ç£ï¼½ã€€å…¥åŠ›ã€€æŠ¼ä¸‹")
			vk_operation = &"VKO_Pressed"
		
		# ä½•ã‹ã‚­ãƒ¼ã‚’é›¢ã—ãŸã¨ã
		elif event.is_released():
			#print("ï¼»ç›£ç£ï¼½ã€€å…¥åŠ›ã€€ãƒªãƒªãƒ¼ã‚¹")
			vk_operation = &"VKO_Released"
		
		# ãã‚Œä»¥å¤–ã«ã¯å¯¾å¿œã—ã¦ãªã„
		else:
			print("ï¼»ç›£ç£ï¼½ã€€å…¥åŠ›ã€€â–²ï¼æƒ³å®šå¤–")
			return

		# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
		# ä»¥ä¸‹ã€ä»®æƒ³ã‚­ãƒ¼
		# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
		# æ–‡å­—åˆ—ã ã‘ã ã¨ã€æŠ¼ã—ãŸã®ã‹æ”¾ã—ãŸã®ã‹åˆ†ã‹ã‚‰ãªã„
		var event_as_text = event.as_text()
		
		# æ–‡å­—åˆ—ã‚’ãƒœã‚¿ãƒ³ç•ªå·ã«å¤‰æ›
		var button_number = self.monkey().key_config_island().get_button_number_by_text(event_as_text)
		
		# ãƒœã‚¿ãƒ³ç•ªå·ã‚’ã€ä»®æƒ³ã‚­ãƒ¼åã«å¤‰æ›
		var virtual_key_name = self.monkey().key_config_island().get_virtual_key_name_by_button_number(button_number)

		# ãƒ¬ãƒãƒ¼å€¤
		var lever_value = self.monkey().key_config_island().get_lever_value_by_text(event_as_text)

		# ä»®æƒ³ã‚­ãƒ¼ã‚’æŠ¼ä¸‹ã—ãŸã¨ã„ã†å»ºä»˜ã‘
		self.on_virtual_key_input(virtual_key_name, lever_value, vk_operation)


# ä»®æƒ³ã‚­ãƒ¼ã‚’æŠ¼ä¸‹ã—ãŸã¨ã„ã†å»ºä»˜ã‘
#
# å‘¼å‡ºã—å…ƒ:
# 	_unhandled_key_input()
#	_unhandled_input()
func on_virtual_key_input(virtual_key, lever_value, vk_operation):

	# ç¾åœ¨ã®ãƒ‡ãƒ‘ãƒ¼ãƒˆãƒ¡ãƒ³ãƒˆã«ç´ã¥ãã€é …ç›®ã¯è¾æ›¸ã«è¨˜è¼‰ã•ã‚Œã¦ã„ã‚‹ã‹ï¼Ÿ
	if self.monkey().scenario_writer().on_virtual_key_input(
			virtual_key,
			lever_value,
			vk_operation):
		# å…¥åŠ›ã•ã‚ŒãŸã‚­ãƒ¼ã¸ã®å¯¾å‡¦ãŒå®Œäº†ã—ãŸãªã‚‰ã€å‡¦ç†ã‚’æŠœã‘ã‚‹
		return

	# ã‚·ãƒŠãƒªã‚ªãƒ©ã‚¤ã‚¿ãƒ¼ãƒ»ãƒãƒ–ã§ã€€ã“ã®å…¥åŠ›ã‚’ã‚¹ãƒ«ãƒ¼ã—ãŸãªã‚‰ã€ä»¥é™ã®å‡¦ç†ã‚’ç¶šã‘ã‚‹
	print("ï¼»ç›£ç£ï¼½ã€€ä»®æƒ³ã‚­ãƒ¼ï¼ˆ" + virtual_key + "ï¼‰ã€€ãƒ¬ãƒãƒ¼å€¤ï¼š" + str(lever_value) + "ã€€æ“ä½œï¼š" + vk_operation)

	# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ãƒ»ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ã¸æ¸¡ã™
	self.monkey().programmer().scenario_player().get_current_message_window_gui().on_virtual_key_input(virtual_key, lever_value, vk_operation)
